mining: add applySolution() to interface

Sjors opened the pull request:
Unlike the `submitblock` RPC which takes a fully serialized block, when a block solution is received via IPC the client only provides the nonce, coinbase and a few other fields. It may not have all the information it needs to reconstruct the block. This makes debugging difficult when the block is invalid.

This PR adds `applySolution()` which returns the reconstructed block and can be used by the client for debugging. It's assigned `@11` in the `.cap` file, and will not break existing client software.

It can also be used by the client for an alternative broadcast mechanism.

The tests cover both mainnet (unit tests) and regtest (functional tests), because the latter has SegWit active.

The second commit documents the fact that `applySolution()` as well as the (already existing) `submitSolution` are slightly different from the `submitsolution` RPC in that they expect a complete coinbase transaction and will not automatically add a witness. This behavior is also covered in the functional test of the first commit.

This is alternative or complimentary approach to:
- #33372

Sjors committed:
mining: add applySolution() to interface

Unlike the submitblock RPC which takes a fully serialized block,
when a block solution is received via IPC the client only provides
the nonce, coinbase and a few other fields. It may not have all
the information it needs to reconstruct the block. This makes
debugging difficult when the block is invalid.

This commit adds applySolution() which returns the reconstructed block
and can be used by the client for debugging.

It can also be used by the client for an alternative broadcast mechanism.

ajtowns commented on src/node/interfaces.cpp:
```diff
@@ -916,6 +916,12 @@ class BlockTemplateImpl : public BlockTemplate
 
+    CBlock applySolution(uint32_t version, uint32_t timestamp, uint32_t nonce, CTransactionRef coinbase) override
+    {
+        AddMerkleRootAndCoinbase(m_block_template->block, std::move(coinbase), version, timestamp, nonce);
```
I think this assumes that the provided coinbase tx includes the coinbase witness, which is different to how `submitblock` behaves. (In particular `submitblock` includes a call to `chainman.UpdateUncommittedBlockStructures`, whereas `AddMerkleRootAndCoinbase` doesn't) That seems like it might be a source of bugs -- might at least be worth adding a check and having an explicit log message if the coinbase requires a witness but doesn't have it?

FWIW I think `miner_tests` is being run against mainnet params for ~100 blocks, which means that segwit is never active (it activates at block 481824), so the witness-specific code isn't being tested there.

ryanofsky replied:
> Another thing that's not great here is that `m_block_template->block` is mutated.

It does seem nicer to not to mutate the block template. And as long as entire block is being serialized and sent over the socket anyway, probably the cost of copying it is not significant.

But one thought is that maybe serializing whole block is not necessary, and this could just return CBlockHeader?

Another idea is that this could be changed to use output parameters, and the client could decide what information it wants returned. For example:

```c++
virtual void applySolution(
    uint32_t version, uint32_t timestamp, uint32_t nonce, CTransactionRef coinbase,
    CBlockHeader* out_header=nullptr, CBlock* out_block=nullptr, std::string* out_invalid_reason=nullptr) = 0;
```

The cap'n proto declaration corresponding to this would look like:

```capnp
applySolution @11 (context: Proxy.Context, version: UInt32, timestamp: UInt32, nonce: UInt32, coinbase :Data, wantHeader: Bool, wantBlock :Bool, wantInvalidReason :Bool) 
    -> (header: Data, block: Data, invalidReason: Text);
```

(libmultimultiprocess uses `wantFoo` parameters and `foo` return values to support c++ methods with optional output parameters.)

The idea with the `out_invalid_reason` parameter would be to provide the debug information #33372 was previously logging.

Not sure if this interface would actually be more useful in practice, but wanted to point out it is possible to return information selectively like this.

ajtowns reviewed (commented):
I like this better than just logging stuff, for whatever that's worth.

ryanofsky commented on src/ipc/capnp/mining.capnp:
```diff
@@ -32,6 +32,7 @@ interface BlockTemplate $Proxy.wrap("interfaces::BlockTemplate") {
     getWitnessCommitmentIndex @7 (context: Proxy.Context) -> (result: Int32);
     getCoinbaseMerklePath @8 (context: Proxy.Context) -> (result: List(Data));
     submitSolution @9 (context: Proxy.Context, version: UInt32, timestamp: UInt32, nonce: UInt32, coinbase :Data) -> (result: Bool);
+    applySolution @11 (context: Proxy.Context, version: UInt32, timestamp: UInt32, nonce: UInt32, coinbase :Data) -> (result: Data);
```
re: https://github.com/bitcoin/bitcoin/pull/33374#discussion_r2348144204

I think it makes sense if method order order in `.capnp` files just matches the method order in `.h` files so the declarations are easy to compare.

Since we haven't needed binary compatibility before, we've just renumbered when making any changes, so numbers have been in order up to this point. But if we want to start having binary compatibility, we won't be able to reassign numbers anymore, and it'll be normal for them not to be in order in the file.

> * = it's still not a big deal at this stage to break clients

Makes sense, but to be sure, there are different ways of breaking ciients. For example if we just make changes like renumbering mining methods, or fields, or changing default values, results could be a little chaotic if an old client connects to a new node or vice versa, because client IPC calls could appear to work but invoke the wrong methods and send invalid parameter values.

A way to make cleaner breaks is to update the `Init.makeMining` method. Give it a new number so only new clients will be able to call it, and add a `Init.makeMiningRemoved` or `init.makeMiningV1` method with the old number that returns null or triggers an error. This way an old client connecting to a new node, or new cient connecting to an old node will fail early instead of sending requests later that would not be interpreted correctly:

```diff
--- a/src/ipc/capnp/init.capnp
+++ b/src/ipc/capnp/init.capnp
@@ -19,5 +19,8 @@ using Mining = import "mining.capnp";
 interface Init $Proxy.wrap("interfaces::Init") {
     construct @0 (threadMap: Proxy.ThreadMap) -> (threadMap :Proxy.ThreadMap);
     makeEcho @1 (context :Proxy.Context) -> (result :Echo.Echo);
-    makeMining @2 (context :Proxy.Context) -> (result :Mining.Mining);
+    makeMining @3 (context :Proxy.Context) -> (result :Mining.Mining);
+
+    # Previous declarations for binary compatibility, no longer supported.
+    makeMiningV1 @2 (context :Proxy.Context) -> (result :Mining.Mining);
 }
--- a/src/interfaces/init.h
+++ b/src/interfaces/init.h
@@ -38,6 +38,9 @@ public:
     virtual std::unique_ptr<Echo> makeEcho() { return nullptr; }
     virtual Ipc* ipc() { return nullptr; }
     virtual bool canListenIpc() { return false; }
+
+    // For IPC compatibility, no longer supported.
+    virtual std::unique_ptr<Mining> makeMiningV1() { return nullptr; }
 };
 
 //! Return implementation of Init interface for the node process. If the argv
```

ryanofsky reviewed (approved):
Concept ACK, this does seem like a more general solution than 33372.

Sjors committed:
doc: submitSolution doesn't add a witness

Sjors commented:
On second thought I don't think mutating the block is a problem.

> I think this assumes that the provided coinbase tx includes the coinbase witness, which is different to how `submitblock` behaves. (In particular `submitblock` includes a call to `chainman.UpdateUncommittedBlockStructures`, whereas `AddMerkleRootAndCoinbase` doesn't) That seems like it might be a source of bugs -- might at least be worth adding a check and having an explicit log message if the coinbase requires a witness but doesn't have it?

IIUC only the `submitblock` **RPC** adds the coinbase witness, whereas `submitSolution()` and `applySolution()` IPC calls both don't.

This has been a source of confusion:
- https://github.com/stratum-mining/sv2-spec/discussions/156
- https://github.com/Sjors/sv2-tp/pull/10

So far the SRI software has been giving us a complete coinbase, including the coinbase witness. But this involved a hardcoded assumption about its value. So the idea was to have the Template Provider communicate this value.

But it seems that approach in RPC land was different: we don't communicate the coinbase witness to pool software and just fix it when they submit the solution. The description of says "This is safe for submitted blocks.", but that's only true if the pool took our SegWit `OP_RETURN`.

The latter is not unreasonable, but I would prefer that we pick one approach:
1. Give pools both the `OP_RETURN` and coinbase witness; or
2. The pool figures out both values

I prefer (1).

cc @plebhash.

> FWIW I think `miner_tests` is being run against mainnet params for ~100 blocks, which means that segwit is never active (it activates at block 481824), so the witness-specific code isn't being tested there.

I rebased this on #33380 and added a test on regtest. This checks that dropping the coinbase witness has an effect. You can verify this test by modifying `applySolution()`:

```cpp
AddMerkleRootAndCoinbase(m_block_template->block, std::move(coinbase), version, timestamp, nonce);
LOCK(cs_main);
# yolo
chainman().UpdateUncommittedBlockStructures(m_block_template->block, chainman().ActiveTip());
return m_block_template->block;
```

plebhash commented:
> The current ambition is to have the Template Provider communicate this value (eventually, no rush).

this would require a change to the Sv2 spec so that `NewTemplate` message carries a new field called `coinbase_witness `. 

but unfortunately, changing Sv2 spec is usually received with resistance from the current players.

if/when a softfork happens and we start committing stuff to the coinbases' `witness reserved value`, that could be stronger motivation.

but until that day comes, this spec change would be merely motivated for making Sv2 "future-proof" with regards to this hypothetical soft fork.

so this is all to put some extra emphasis on the **(eventually, no rush)** quoted above.

---

> 1. Give pools both the OP_RETURN and coinbase witness; or
> 2. The pool figures out both values
>
> I prefer (1).

In Sv2 terms, option 2. would incur in extra round trips (`RequestTransactionData/.Success`) and computation of witness root, which would translate into undesired delays.

So I'd prefer 1 as well. Even if Bitcoin Core continues to redundantly provide `0x000...000` as coinbase witness and Sv2 spec continues sticking to the assumption that coinbase witness always has this same value.

So whenever we get consensus a Sv2 spec change (either preemptively or forced by a soft-fork), the Sv2 TP would already have means to fill `NewTemplate.coinbase_witness` with appropriate values coming from Bitcoin Core via IPC.

---

But AFAIU, that's somewhat different from the proposal on this PR, right?

IIUC this PR is only adding a new `applySolution`, and the discussion point now is simply the assumptions about how such solution would be crafted by Sv2 software.

Sjors commented:
@plebhash indeed this PR just introduces the extra method and clarifies in the documentation that `submitSolution` and `applySolution` are slightly different from the `submitsolution` RPC.

Sjors commented:
> 1. Main question is why is this new applySolution method needed at all if `submitSolution()` modifies the block and there is already a `getBlock()` method?

This is a great point. Let's just embrace the fact that `submitSolution()` modifies the block.

I'll extract some of the documentation improvements, and add some based on your comments, and PR those later.

Sjors commented on src/test/miner_tests.cpp:
```diff
@@ -758,7 +758,15 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
             BOOST_REQUIRE(Assert(m_node.chainman)->ProcessNewBlock(shared_pblock, /*force_processing=*/true, /*min_pow_checked=*/true, nullptr));
         } else {
-            BOOST_REQUIRE(block_template->submitSolution(block.nVersion, block.nTime, block.nNonce, MakeTransactionRef(txCoinbase)));
+            const CTransactionRef coinbase{MakeTransactionRef(txCoinbase)};
+            BOOST_REQUIRE(block_template->submitSolution(block.nVersion, block.nTime, block.nNonce, coinbase));
```
This test only calls `submitSolution()` once, unless I'm missing something?
